core/src/actors/session.rs

```diff
@@ -8,73 +7,123 @@ use actix::{
     Actor, ActorContext, ActorFuture, AsyncContext, Context, ContextFutureSpawner, Handler,
     Message, Running, StreamHandler, System, WrapFuture,
 };
-use log::debug;
+use log::{debug, info, warn};
 use tokio::io::WriteHalf;
 use tokio::net::TcpStream;
 
-use crate::actors::codec::{P2PCodec, Request, Response};
+use crate::actors::codec::{BytesMut, P2PCodec};
 use crate::actors::peers_manager;
-use crate::actors::sessions_manager::{Register, SessionsManager, Unregister};
+use crate::actors::sessions_manager::{Consolidate, Register, SessionsManager, Unregister};
 
 use witnet_data_structures::{
+    builders::from_address,
     serializers::TryFrom,
-    types::{Command, Message as WitnetMessage},
+    types::{Address, Command, Message as WitnetMessage},
 };
 use witnet_p2p::sessions::{SessionStatus, SessionType};
 
 ////////////////////////////////////////////////////////////////////////////////////////
 // ACTOR BASIC STRUCTURE
 ////////////////////////////////////////////////////////////////////////////////////////
+/// Handshake flags
+#[derive(Default)]
+struct HandshakeFlags {
+    /// Flag to indicate that a version message was tx
+    version_tx: bool,
+    /// Flag to indicate that a version message was rx
+    version_rx: bool,
+    /// Flag to indicate that a verack message was tx
+    verack_tx: bool,
+    /// Flag to indicate that a verack message was rx
+    verack_rx: bool,
+}
+
+/// HandshakeFlags helper methods
+impl HandshakeFlags {
+    // Auxiliary function to check if all flags are set to true
+    fn all_true(&self) -> bool {
+        self.version_tx && self.version_rx && self.verack_tx && self.verack_rx
```

Comment:
Make sure this flags are ordered putting _the most likely to be false_ in front at the time of evaluation.
`</performanceParanoia>`